src/validation/difficulty.cairo

```diff
@@ -71,37 +71,35 @@ fn target_to_bits(target: u256) -> Result<u32, ByteArray> {
         return Result::Err("Exceeds max value");
     }
 
-    // Find the most significant byte
-    let mut size: u32 = 32;
-    let mut compact = target;
-
-    // Count leading zero bytes by finding the first non-zero byte
-    while size > 1 && shr(compact, (size - 1) * 8) == 0 {
-        size -= 1;
+    let mut divider = MAX_TARGET;
+    let mut tmp = target;
+    let mut padding = 3;
+    let mut last_coef = 0;
+
+    // Get bytes from most to least significant byte
+    for _ in 0
+        ..31_u8 {
+            let (q, r) = core::traits::DivRem::div_rem(tmp, divider.try_into().unwrap());
+            divider /= 0x100;
+            last_coef = tmp;
+            tmp = r;
+            if (q != 0) {
+                // Normalize
+                if (q > 0x80) {
+                    padding -= 1;
+                }
+                break;
+            }
+            padding += 1;
+        };
+
+    // 29 = 32 - 3(coefficient bytes)
+    for _ in 0..(29 - padding) {
+        last_coef /= 0x100;
     };
 
-    // Extract mantissa (most significant 3 bytes)
-    let mut mantissa: u32 = shr(compact, (size - 3) * 8).try_into().unwrap();
-
-    // Normalize
-    if mantissa > 0x7fffff {
-        mantissa = (mantissa + 0x80) / 0x100;
-        size += 1;
-    }
-
-    // Ensure the mantissa is only 3 bytes
-    mantissa = mantissa & 0xffffff;
-
-    // Check size doesn't exceed maximum
-    if size > 34 {
-        return Result::Err("Overflow");
-    }
-
-    // Convert size to u256
-    let size_u256: u256 = size.into();
-
-    // Combine size and mantissa
-    let result: u32 = (shl(size_u256, 24_u32) + mantissa.into()).try_into().unwrap();
+    // Combine exponent and coefficient
+    let result: u32 = (shl((32 - padding), 24_u32) + last_coef.into()).try_into().unwrap();
```

Comment:
Can we multiply by a const 2^25 here instead of shl?

Reply:
Yes right !

Reply:
Maybe we can get rid of shifts completely by replacing them with static lookups instead? Since we always know the max shift size.

```
if shl == 1 { return val * 2 } 
else if shl == 2 { return val * 4 }
...
```

Reply:
yes, for the times when the shift left is dynamic I'll explore that, but here it's easy, it's fixed.

Reply:
you meen something like this
```
fn shr(value: u64, shift: u64) -> u64 {
    return match shift {
        0 => value,
        1 => value / 2,
        2 => value / 4,
        3 => value / 8,
        4 => value / 16,
```
In the case of fn compute_block_reward, there are 64 possibilities

Reply:
I suspect that match is just a sugar over a series of ifs. It probably will be faster to put pow results into a const array:
```
const pow: [u64, 64] = [1, 2, 4, 8, ...];
...
return value / pow[shift]
```

Reply:
In general I would try to optimize shr/shl this way before replacing them with manual div/mul which makes code hardly readable.

Reply:
If think match is more optimized than succesives if / else. 
The reason match only allow values to succesive and in order, from 0,1,2,3,4,5 is that the low level cairo can probably jump on the case, whereas on successive ifs it will check all of them.

Reply:
Test and confirm the above.

Reply:
Actually, if the" _" is also a number (like => 0) instead of a panic, the match pattern gives a constant access cost that always beat the if statement except for i=0.

Reply:
See https://github.com/feltroidprime/cairo-perfs-snippets/blob/main/src/pow_2_lookup.cairo